src/_pytest/tmpdir.py

```diff
@@ -2,21 +2,116 @@
 from __future__ import absolute_import, division, print_function
 
 import re
+from six.moves import reduce
 
 import pytest
 import py
 from _pytest.monkeypatch import MonkeyPatch
+from .compat import Path
+import attr
+import shutil
+import tempfile
 
 
+def make_numbered_dir(root, prefix):
+    l_prefix = prefix.lower()
+
+    def parse_num(p, cut=len(l_prefix)):
+        maybe_num = p.name[cut:]
+        try:
+            return int(maybe_num)
+        except ValueError:
+            return -1
+
+    for i in range(10):
+        # try up to 10 times to create the folder
+        max_existing = reduce(
+            max,
+            (
+                parse_num(x)
+                for x in root.iterdir()
+                if x.name.lower().startswith(l_prefix)
+            ),
+            -1,
+        )
```

Comment:
this code was a major head scratcher for me, maybe refactor it to not use `reduce`?

Reply:
there is not `max(iterable, default=...)` on python2

Reply:
i extracted it a bit - the for loop is slower due to executing on the interpreter, impending push after i have a first iteration of folder cleanup